src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1944,10 +1971,173 @@ private class OffsetAndRule {
+                {
+                    id = FindTimeZoneId(rawData);
+                }
+
+                return rawData != null;
+            }
+            catch (IOException)
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+        }
+
+        private static string GetTzEnvironmentVariable()
+        {
+            string result = Environment.GetEnvironmentVariable("TZ");
+            if (!string.IsNullOrEmpty(result))
+            {
+                if (result[0] == ':')
+                {
+                    // strip off the ':' prefix
+                    result = result.Substring(1);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Find the time zone id by searching all the tzfiles for the one that matches rawData
+        /// and return its file name.
+        /// </summary>
+        private static string FindTimeZoneId(byte[] rawData)
+        {
+            string id = null;
+            string localtimeFilePath = Path.Combine(c_timeZoneDirectory, "localtime");
+
+            foreach (string filePath in Directory.EnumerateFiles(c_timeZoneDirectory, "*", SearchOption.AllDirectories))
+            {
+                // skip the localtime file, since that won't give us the correct id
+                if (filePath != localtimeFilePath)
+                {
+                    try
+                    {
+                        using (FileStream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+                        {
+                            int fileLength = (int)stream.Length;
+                            if (fileLength == rawData.Length)
+                            {
+                                int index = 0;
+                                int count = fileLength;
+                                byte[] bytes = new byte[count];
+                                while (count > 0)
+                                {
+                                    int n = stream.Read(bytes, index, count);
+                                    if (n == 0)
+                                        __Error.EndOfFile();
+                                    index += n;
+                                    count -= n;
+                                }
+
+                                for (int i = 0; i<fileLength; i++)
+                                {
+                                    if (bytes[i] != rawData[i])
+                                    {
+                                        continue;
+                                    }
+                                }
+
+                                // if all bytes are the same, this must be the right tz file
+                                id = filePath;
+
+                                // strip off the root time zone directory
+                                string timeZoneDirectory = c_timeZoneDirectory + "/";
```

Comment:
Earlier I suggested just making c_timeZoneDirectory end with a '/', in which case you won't need this concatenation.